src/test/java/hudson/plugins/mercurial/MercurialSCMBuilderTest.java

```diff
@@ -0,0 +1,88 @@
+package hudson.plugins.mercurial;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import hudson.FilePath;
+import hudson.tools.ToolProperty;
+import hudson.util.LogTaskListener;
+import hudson.util.StreamTaskListener;
+import jenkins.scm.api.SCMHead;
+import jenkins.scm.api.SCMHeadObserver;
+import jenkins.scm.api.SCMRevision;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+import org.jvnet.hudson.test.JenkinsRule;
+
+import static org.junit.Assert.assertEquals;
+
+public class MercurialSCMBuilderTest {
+    @Rule
+    public JenkinsRule j = new JenkinsRule();
+    @Rule
+    public MercurialRule m = new MercurialRule(j);
+    @Rule
+    public TemporaryFolder tmp = new TemporaryFolder();
+
+    private static LogTaskListener listener;
+    private static MercurialSCMSource mercurialSCMSource;
+
+    @Before
+    public void prepareEnvironment() throws Exception {
+        String instName = "caching";
+        MercurialInstallation installation = new MercurialInstallation(instName, "", "hg", false, true, null, false, null,
+                Collections.<ToolProperty<?>>emptyList());
+        listener = new LogTaskListener(Logger.getLogger(MercurialSCMSourceTest.class.getName()), Level.INFO);
+        j.jenkins.getDescriptorByType(MercurialInstallation.DescriptorImpl.class).setInstallations(installation);
+        FilePath repo = new FilePath(tmp.getRoot());
+        installation.forNode(j.jenkins, StreamTaskListener.fromStdout());
+        m.hg(repo, "init");
+        repo.child("file").write("initial content", "UTF-8");
+        m.hg(repo, "commit", "--addremove", "--message=initial");
+        m.hg(repo, "tag", "version-1.0");
+        m.hg(repo, "branch", "my-branch");
+        repo.child("file2").write("content in branch", "UTF-8");
+        m.hg(repo, "commit", "--addremove", "--message=branch");
+        m.hg(repo, "tag", "version-1.1");
+
+        installation.forNode(j.jenkins, StreamTaskListener.fromStdout());
+        mercurialSCMSource = new MercurialSCMSource(null, instName, tmp.getRoot().toURI().toURL().toString(), null, null, null, null, null, true);
+    }
+
+    @Test
+    public void headNameEquals() throws IOException, InterruptedException {
+        Map<SCMHead, SCMRevision> result = mercurialSCMSource.fetch(null, SCMHeadObserver.collect(), null, null).result();
+        for (Map.Entry<SCMHead, SCMRevision> entry : result.entrySet()) {
+            MercurialSCM mercurialSCM = new MercurialSCMBuilder(entry.getKey(), entry.getValue(), "", "")
+                    .build();
+            assertEquals(MercurialSCM.RevisionType.CHANGESET, mercurialSCM.getRevisionType());
+            assertEquals(entry.getValue().getHead().getName(), mercurialSCM.getHeadName());
```

Comment:
This test is just asserting that a newly introduced method has an expected value. Fine so far as it goes, but does not demonstrate that the purported purpose of the change (to fix some behavior of `resolveScm`) is actually accomplished by doing that. Is some piece of code using reflection to look for a JavaBeans property named `headName`? (I certainly hope not!)

Reply:
no reflection here of course) just check all existing heads) `resolveScm` works correct just it return `MercurialSCM` where you can get only `hash` of commit. in fact i dont know which test need for it but my tests confirm what `headName` correct. maybe need other name of variable) but it's not breaking change like change value of `revision` from `hash` to `branch name`